crates/telemetry-subscribers/src/span_latency_prom.rs

```diff
@@ -0,0 +1,102 @@
+//! Thus this is a much smaller and more focused module.
+//!
+//! ## Making spans visible
+//! This module can only record latencies for spans that get created.  By default, this is controlled by
+//! env_filter and logging levels.
+
+use chrono::offset::Utc;
+use prometheus::{exponential_buckets, register_histogram_vec_with_registry, Registry};
+use tracing::{span, Subscriber};
+
+/// A tokio_tracing Layer that records span latencies into Prometheus histograms
+pub struct PrometheusSpanLatencyLayer {
+    span_latencies: prometheus::HistogramVec,
+}
+
+#[derive(Debug)]
+pub enum PrometheusSpanError {
+    /// num_buckets must be positive >= 1
+    ZeroOrNegativeNumBuckets,
+    PromError(prometheus::Error),
+}
+
+impl From<prometheus::Error> for PrometheusSpanError {
+    fn from(err: prometheus::Error) -> Self {
+        Self::PromError(err)
+    }
+}
+
+const TOP_LATENCY_IN_NS: f64 = 300.0 * 1.0e9;
+const LOWEST_LATENCY_IN_NS: f64 = 500.0;
+
+impl PrometheusSpanLatencyLayer {
+    /// Create a new layer, injecting latencies into the given registry.
+    /// The num_buckets controls how many buckets thus how much memory and time series one
+    /// uses up in Prometheus (and in the application).  10 is probably a minimum.
+    pub fn try_new(registry: &Registry, num_buckets: usize) -> Result<Self, PrometheusSpanError> {
+        if num_buckets < 1 {
+            return Err(PrometheusSpanError::ZeroOrNegativeNumBuckets);
+        }
+
+        // Histogram for span latencies must accommodate a wide range of possible latencies, so
+        // don't use the default Prometheus buckets
+        // The latencies are in nanoseconds
+        let factor = (TOP_LATENCY_IN_NS / LOWEST_LATENCY_IN_NS).powf(1.0 / (num_buckets as f64));
+        let buckets = exponential_buckets(LOWEST_LATENCY_IN_NS, factor, num_buckets)?;
+        let span_latencies = register_histogram_vec_with_registry!(
+            "tracing_span_latencies",
+            "Latencies from tokio-tracing spans",
+            &["span_name"],
+            buckets,
+            registry
+        )?;
+        Ok(Self { span_latencies })
+    }
+}
+
+struct PromSpanTimestamp(i64);
+
+impl<S> tracing_subscriber::Layer<S> for PrometheusSpanLatencyLayer
+where
+    S: Subscriber + for<'span> tracing_subscriber::registry::LookupSpan<'span>,
+{
+    fn on_new_span(
+        &self,
+        _attrs: &span::Attributes,
+        id: &span::Id,
+        ctx: tracing_subscriber::layer::Context<S>,
+    ) {
+        let span = ctx.span(id).unwrap();
+        // NOTE: there are other extensions that insert timings.  For example,
+        // tracing_subscriber's with_span_events() inserts events at open and close that contain timings.
+        // However, we cannot be guaranteed that those events would be turned on.
+        // TODO: maybe add a check if something else is already inserted...
+        span.extensions_mut()
+            .insert(PromSpanTimestamp(Utc::now().timestamp()));
```

Comment:
`Instant::now` rather then `Utc` is more better fit for local time measurement 
- it is way faster: Instance::now is a register read while Utc::now is often a syscall
- it is also more reliable: local clock returned by Utc can go back and forth while Instant is monotonic

Reply:
Ok, sounds good